### PR TITLE
WIP: SQL: Implement TIME_PARSE function for parsing strings

### LINKED ISSUE #54963: SQL: Implement TIME_PARSE
Implement TIME_PARSE(<time_str>, <pattern_str>) function
which allows to parse a time string according to the specified
pattern into a time object. The patterns allowed are those of
java.time.format.DateTimeFormatter.

Follows: #54960

### PR DESCRIPTION
**Pull request by vishnugt**
Implement TIME_PARSE(<time_str>, <pattern_str>) function
which allows to parse a time string according to the specified
pattern into a time object. The patterns allowed are those of
`java.time.format.DateTimeFormatter`.

This closes #54963

### CODE REVIEW

**Comment by vishnugt** [comment]
Should we limit the TIME_PARSE function only to <time_str> or can we allow the users to give <datetime_str> as well?

I've yet to add test cases and docs for this, but would you mind taking a look at the core part?
@matriv

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/TimeParseProcessor.java [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+            throw new SqlIllegalArgumentException("A string is required; received [{}]", pattern);
+        }
+
+        if (((String) timestampStr).isEmpty() || ((String) pattern).isEmpty()) {
+            return null;
+        }
+
+        try {
+            TemporalAccessor ta = DateTimeFormatter.ofPattern((String) pattern, Locale.ROOT)
+                .parseBest((String) timestampStr, ZonedDateTime::from, LocalTime::from);
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/TimeParseProcessor.java:72]:
```suggestion
                .parseBest((String) timestampStr, OffsetTime::from, LocalTime::from);
```

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/TimeParseProcessor.java [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+            return ta;
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/TimeParseProcessor.java:73]: If it's parsed as LocalTime then you need to apply `UTC` timezone to make it `OffsetTime`

File: x-pack/plugin/sql/qa/src/main/resources/command.csv-spec [hunk 3]
```diff
@@ -46,6 +46,7 @@ DATEDIFF         |SCALAR
 DATEPART         |SCALAR
 DATETIME_FORMAT  |SCALAR
 DATETIME_PARSE   |SCALAR
+TIME_PARSE       |SCALAR
```
Comment (matriv) [inline x-pack/plugin/sql/qa/src/main/resources/command.csv-spec:49]: Please use alphanumeric ordering within the date/time functions.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/SqlFunctionRegistry.java [hunk 4]
```diff
@@ -172,6 +173,7 @@ public SqlFunctionRegistry() {
                 def(DatePart.class, DatePart::new, "DATEPART", "DATE_PART"),
                 def(DateTimeFormat.class, DateTimeFormat::new, "DATETIME_FORMAT"),
                 def(DateTimeParse.class, DateTimeParse::new, "DATETIME_PARSE"),
+                def(TimeParse.class, TimeParse::new, "TIME_PARSE"),
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/SqlFunctionRegistry.java:176]: same here

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/Processors.java [hunk 5]
```diff
@@ -87,6 +88,7 @@ private Processors() {}
         entries.add(new Entry(Processor.class, DatePartProcessor.NAME, DatePartProcessor::new));
         entries.add(new Entry(Processor.class, DateTimeFormatProcessor.NAME, DateTimeFormatProcessor::new));
         entries.add(new Entry(Processor.class, DateTimeParseProcessor.NAME, DateTimeParseProcessor::new));
+        entries.add(new Entry(Processor.class, TimeParseProcessor.NAME, TimeParseProcessor::new));
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/Processors.java:91]: same here

**Review by matriv — COMMENTED** [review]
Left some comments.

Please try to follow an approach like here: https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateTimeProcessor.java#L21 so there is an enum used and there is no need for new Pipe, and Processor class.

Just a reminder, don't forget to add docs, doc examples/tests, integration tests, tests in VerifierErrorMessageTests for the validation of the argument date types, you can check the PR that implemented the DATETIME_PARSE for reference.

Thank you!

**Comment by matriv** [comment]
> Should we limit the TIME_PARSE function only to <time_str> or can we allow the users to give <datetime_str> as well?

No, it should only accept time only (possibly with tz or offset info), as the return data type of the function should be `TIME` only, so any other combination of str + pattern should throw an exception (time parts only or both date and time units).

> I've yet to add test cases and docs for this, but would you mind taking a look at the core part?

Left just a reminder for this.

**Comment by matriv** [comment]
@vishnugt FYI: you can run `./gradlew --parallel -p x-pack/plugin/sql check` locally to check that everything is ok.

**Comment by vishnugt** [comment]
Thanks for the review, will make the suggested changes, and add docs and test cases and will ping here! 

> No, it should only accept time only (possibly with tz or offset info), as the return data type of the function should be TIME only, so any other combination of str + pattern should throw an exception (time parts only or both date and time units).

> (time parts only or both date and time units).
You meant to say "date parts only or both date and time units" right?

**Comment by matriv** [comment]
> You meant to say "(date parts only or both date and time units)" right?

Yes, sorry for confusion.

**Comment by vishnugt** [comment]
Hey, sorry for the trouble.  Due to some stuff, I won't be able to able to continue working on this PR.  Sorry for the trouble again.

**Comment by matriv** [comment]
@vishnugt No worries, Thank you for informing, hope everything is ok.

**Comment by matriv** [comment]
@Patrick0308

Would you like to take over?

**Comment by Patrick0308** [comment]
@matriv OK, Should I reopen PR？

**Comment by matriv** [comment]
@Patrick0308 Yes Please!

**Comment by matriv** [comment]
Closing as it's superseded by #55095

**Pull request closed without merging**